cdap-app-fabric/src/main/java/io/cdap/cdap/app/guice/DistributedProgramContainerModule.java

```diff
@@ -292,4 +298,63 @@ protected void configure() {
   private static String generateClientId(ProgramRunId programRunId, String instanceId) {
     return String.format("%s.%s.%s", programRunId.getParent(), programRunId.getRun(), instanceId);
   }
+
+  /**
+   * A Guice module to provide {@link ProgramStateWriter} binding. In normal same cluster / remote cluster execution,
+   * it just publishes program states to TMS based on the discovery service. For tethered execution,
+   * it publishes program states to the TMS running in the current master environment,
+   * instead of using the normal discovery service that bind to talk to the originating CDAP instance.
+   */
+  private static final class ProgramStateWriterModule extends PrivateModule {
```

Comment:
Does this need to be a private module because you don't want to expose the MasterEnvironment instance binding?

Reply:
Yes. The principle of minimum exposure.